### PR TITLE
add persistence to JSON file for flashcards in hidden directory

### PR DESCRIPTION
**Pull request by chuyouchia**
output: 

image

Process:

https://github.com/reorproject/reor/assets/58682208/2199c2bc-1e43-4207-aad3-981eeec2a747

### CODE REVIEW

File: src/components/Chat/Chat.tsx [hunk 1]
```diff
@@ -258,6 +267,22 @@ const ChatWithLLM: React.FC<ChatWithLLMProps> = ({
     fetchDefaultModel();
   };
 
+  const parseFlashcardAnswer = (
+    messageToBeParsed: string
+  ): { question: string; answer: string }[] => {
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    const [actualOutput, _fileReferences] = messageToBeParsed.split(
```
Comment (chuyouchia) [inline src/components/Chat/Chat.tsx:274]: this should be safe, even if the output doesn't contain file delimiter because it will by default return original output

File: src/components/Chat/Chat.tsx [hunk 2]
```diff
@@ -362,6 +387,32 @@ const ChatWithLLM: React.FC<ChatWithLLMProps> = ({
...
+                const trimmedFileName = removeFileExtension(fileName);
+                await window.files.writeFile({
+                  filePath: `${window.electronStore.getUserDirectory()}/.flashcards/${trimmedFileName}.json`,
+                  content: JSON.stringify(
+                    {
+                      fileGeneratedFrom: currentFilePath,
+                      qnaPairs: parseFlashcardAnswer(currentBotMessage.content),
+                    },
+                    null,
+                    4
```
Comment (chuyouchia) [inline src/components/Chat/Chat.tsx:408]: This uses native pretty printing of JSON

**Comment by samlhuillier** [comment]
This seems to contain a lot of work to parse the generated flashcard strings in the right format. Shouldn't we just use function calling and JSON mode (from Ollama) to generate the json directly? That should give us better performance and better adherence to the types

File: src/components/Chat/Chat.tsx [hunk 3]
```diff
@@ -362,6 +391,32 @@ const ChatWithLLM: React.FC<ChatWithLLMProps> = ({
...
+          currentBotMessage?.content.includes(ANSWER_FORMAT) && (
+            <ChatAction
+              actionText="Convert to flashcard"
+              onClick={async () => {
+                const fileName = await window.path.basename(
+                  currentFilePath || ""
+                );
+                const trimmedFileName = removeFileExtension(fileName);
+                await window.files.writeFile({
+                  filePath: `${window.electronStore.getUserDirectory()}/.flashcards/${trimmedFileName}.json`,
```
Comment (samlhuillier) [inline src/components/Chat/Chat.tsx:405]: Minor issue here with paths for windows - why now use ```window.path.join```?

**Review by samlhuillier — CHANGES_REQUESTED** [review]

**Comment by samlhuillier** [comment]
I'm not too keen on the amount of custom logic for flashcards there is now in the chat component. Ideally in the follow on PRs, in addition to making the component to interact with generated flashcards, we should also have the logic for generating flashcards there and not in Chat.tsx.

**Comment by chuyouchia** [comment]
> I'm not too keen on the amount of custom logic for flashcards there is now in the chat component. Ideally in the follow on PRs, in addition to making the component to interact with generated flashcards, we should also have the logic for generating flashcards there and not in Chat.tsx.

Yeah we should definitely separate them out, so that chat can be more pure and focused.

**Comment by chuyouchia** [comment]
> This seems to contain a lot of work to parse the generated flashcard strings in the right format. Shouldn't we just use function calling and JSON mode (from Ollama) to generate the json directly? That should give us better performance and better adherence to the types

Documenting our discussion here - Streaming and JSON mode is not very compatible, because the information comes in bits and pieces, it is hard to parse it properly for a chat message compatible format. However, function calling is going to be huge in other tasks that we delegate to our chat agent to do

**Review by samlhuillier — APPROVED** [review]
lgtm
